src/library_async.js

```diff
@@ -671,7 +672,14 @@ mergeInto(LibraryManager.library, {
 #if ASYNCIFY_DEBUG
           err('ASYNCIFY: start: ' + start);
 #endif
-          Module['asm'][start]();
+          var asyncReturnValue = Module['asm'][start]();
```

Comment:
`Module['asm'][start]()` returns the return value on it's final call once, all asynchronous execution has completed.

Reply:
It's unclear to me how asyncReturnValue differs from returnValue from earlier above, which is returned from handleSleep?

Reply:
I believe it is a different return value. I tried returning that one and it always seemed to hold the value `0` (its default value). It seems to be for this return value: https://emscripten.org/docs/porting/asyncify.html#returning-values

i.e. `Asyncify.returnValue` holds the optional return value you return from `asynchronous JS ----> C` using `wakeUp(<contents of returnValue>)`. Then, `Asyncify.returnValue` becomes the return value of the C function in [`ASYNCIFY_IMPORTS`](https://emscripten.org/docs/porting/asyncify.html#more-on-asyncify-imports).  For example on [this example](https://emscripten.org/docs/porting/asyncify.html#returning-values), it becomes the return value of the C function `get_digest_size()`. Eventually, `Asyncify.returnValue` becomes the return value of the **last** C function in [`ASYNCIFY_IMPORTS`](https://emscripten.org/docs/porting/asyncify.html#more-on-asyncify-imports) to be executed. 

However, this is **not** the same as the return value of the function being `ccall`-ed. The trivial example is the following:

```c
int foo() {
     printf("Before Sleep\n");
     emscripten_sleep(1000);
     printf("After Sleep\n");
     return 1337;
}
```

If `foo` is being `ccall`-ed, we should return `1337`. However, if you return `Asyncify.returnValue` it will be the return value of `emscripten_sleep`, or `0` since it is `void`. This is because `emscripten_sleep` is the function in `ASYNCIFY_IMPORTS` and it is defined like so:
```js
  emscripten_sleep: function(ms) {
    Asyncify.handleSleep(function(wakeUp) {
      Browser.safeSetTimeout(wakeUp, ms);
    });
  },
```
When `setTimeout` calls `wake` with no arguments (`wakeUp(<VOID NO RETURN VALUE>)`), so `Asyncify.returnValue` remains it's old, default value of 0.


On the other hand, `asyncReturnValue` will contain the return value of `foo`. When a WASM function is asyncified, it is called `N` times by `Module['asm'][start]()` to restart the function execution, if there are `N` unwindings of the stack during the C execution. On the final call of `Module['asm'][start]()`, it will actually have executed it's `return` executed and will return its return value from the WASM function the JavaScript code which we capture into `asyncReturnValue` and return to the `asyncFinalizers`.

Reply:
I agree this is confusing though due to this comment:

```js
    // The return value from a synchronous call is stored here,
    // so we can return it later after rewinding finishes.
    returnValue: 0,
```

It is the return value of the synchronous call of the function in `ASYNC_IMPORTS` (`emscripten_sleep`), not the return value of a function that has called `emscripten_sleep` sleep like `foo`.

Reply:
Added comments explaining the differences to the source file, but perhaps could even benefit from renaming of the variables. I'm thinking:

`Asyncify.returnValue` ---> `handleSleepReturnValue`
`asyncReturnValue` ---> `asyncWasmReturnValue`